cpp/src/groupby/sort/group_rank_scan.cu

```diff
@@ -93,53 +123,212 @@ std::unique_ptr<column> rank_generator(column_view const& order_by,
 
   return ranks;
 }
+
+template <typename value_resolver, typename scan_operator>
+std::unique_ptr<column> rank_generator_reverse(column_view const& grouped_values,
```

Comment:
There's a lot of duplication here. Can't the `rank_generator` just have an extra parameter (or template parameter) or two to indicate the iterator to use in the `inclusive_scan_by_key` (might need some slight trickery since in the non-reverse case it's just a transform iterator with the identity, maybe there's a simpler way to express that though) and the condition to pass the resolver?

Reply:
reduced the code, with tparam and IILE.